### ISSUE TITLE
How to auto switch when a proxy fails?

### ISSUE THREAD

**Issue by hillz1** [FelisCatus/SwitchyOmega#578]
I have 5 different socks5 proxies:

192.168.8.1:1080
192.168.8.1:1081
192.168.8.1:1082
192.168.8.1:1083
192.168.8.1:1084

I want to automatically switch to another proxy when one of my proxies is dead, how can I do that?

**Comment by celeron533** [comment]
I think current OS unable to do the 'smart switch' or 'load balancing' between a collection of profiles.

PS. If you are using Shadowsocks, the latest C# client (2.5.6) supports 'high availability' server settings which will automatically select the best server by connection qualities.

**Comment by hillz1** [comment]
Do you mean the current extension ? I'm using google chrome and this extension, this extension is awesome but it's just missing one single feature, that is switching between proxies when one fails, it would be perfect if it had that feature implemented

**Comment by FelisCatus** [comment]
For now you can create a PAC profile instead and use the following script:

```js
function FindProxyForURL(url, host) {
  return "PROXY 192.168.8.1:1080; PROXY 192.168.8.1:1081; PROXY 192.168.8.1:1082; PROXY 192.168.8.1:1083; PROXY 192.168.8.1:1084";
}
```

Configuration interface for fail-over is more complicated and requires some change in PAC generation code. Suggestions on the interface design (e.g. buttons and inputs) for the feature are welcome.

**Comment by hillz1** [comment]
Too bad that script doesn't work, I can't connect to any sites with that. I have some URLs that need to be excluded as well.

**Comment by FelisCatus** [comment]
@hillz1 The script was written for HTTP proxies instead of SOCKS5 proxies, which is the reason. Sorry for that. I've modified the script above to read `SOCKS5` for each proxy entry. Please try again.

In SwitchyOmega, a PAC profile can also be used in Switch Profiles. You can just create a new Switch Profile, and set its default profile to to the PAC one. After that, you are free to add more rules for all URLs that you want to exclude, and point them to Direct Connection.

**Comment by hillz1** [comment]
Thanks that works pretty well now

**Comment by dyingjedi** [comment]
`For now you can create a PAC profile instead and use the following script:

function FindProxyForURL(url, host) {
  return "SOCKS5 192.168.8.1:1080; SOCKS5 192.168.8.1:1081; SOCKS5 192.168.8.1:1082; SOCKS5 192.168.8.1:1083; SOCKS5 192.168.8.1:1084";
}`

So could use this for HTTP instead for my five different  HTTP proxies?

`For now you can create a PAC profile instead and use the following script:

function FindProxyForURL(url, host) {
  return "HTTP 192.168.8.1:1080; HTTP 192.168.8.1:1081; HTTP 192.168.8.1:1082; HTTP 192.168.8.1:1083; HTTP 192.168.8.1:1084";
}`

Im guessing this would be correct. do i save it mypersonalhomeproxies.pac set it up on my gdrive or something as a pointer to pastbin etc?

Justin Cram
DyingJedi

**Comment by FelisCatus** [comment]
PAC scripts are managed in SwitchyOmega by creating PAC Profiles. Use the New Profile entry in the left side menu, and choose PAC profile in the popup. It should create a profile with a large text area for script editing. The script contents are saved in SwitchOmega and synchronized among devices.

Please note that HTTP proxies use `PROXY` keyword instead of `HTTP`. For your example, the script should be:

```js
function FindProxyForURL(url, host) {
  return "PROXY 192.168.8.1:1080; PROXY 192.168.8.1:1081; PROXY 192.168.8.1:1082; PROXY 192.168.8.1:1083; PROXY 192.168.8.1:1084";
}
```

You can visit http://findproxyforurl.com/ or google for instructions if you have any questions regarding writing PAC scripts.

**Comment by FelisCatus** [comment]
Hi everyone. This issue has been blocked on interface design for a long time. Any idea on how you would like to input multiple proxies on the options page?

**Comment by celeron533** [comment]
UI could be designed as: (there are some alignment problem in this mock-up)
![image](https://cloud.githubusercontent.com/assets/3608762/21074057/b17fee14-bf29-11e6-9e93-98b2085dbc29.png)

**Comment by FelisCatus** [comment]
Interesting, but considering we still have a `Show Advanced` table there, wouldn't it be too much to have such add/delete buttons and rows for every scheme?

**Comment by hillz1** [comment]
One day SwitchyOmega crashed on my pc and I have to write all the rules from scratch, it'd be very helpful if we had an automatic backup feature to dropbox.

**Comment by FelisCatus** [comment]
@hillz1 I would appreciate it if you could create another issue for this.

**Comment by celeron533** [comment]
mock up for `Show Advanced`
![image](https://cloud.githubusercontent.com/assets/3608762/21250789/d7ef7814-c382-11e6-9973-a12e1038d6bf.png)

**Comment by FelisCatus** [comment]
Great mockup, thanks. I noticed you are using Excel as a design tool though. Well that's not a bad thing ~~since the whole Metro UI thing could be designed using the same tool.~~  :wink: This definitely gets the idea of multiple proxies right, but the average user could have problems figuring it out that the proxies are used as fail-overs instead of random or sequential. Also, I feel like the interface is a little bit overwhelming since we now have a lot of fields and buttons.

Another option would be a popup, just like what we did with username & password. In that case, we can drop the protocol/server/port columns and just have one button for customizing the proxies. In that popup window, we can explain that those proxies are used in a failover fashion. This allow new users to discover the feature instead of having to find all over the web and end up here.

Mockup for initial interface, where the plus button would open a popup for fail over.

![image](https://cloud.githubusercontent.com/assets/912649/21487226/0cb3e10e-cb92-11e6-94b7-06c3d4a5a7b1.png)

(The popup window would be something like what @celeron533 proposed, but with more explanation. The four schemes have separate popups.)

Mock up for multiple proxies. The text cannot be edited, and the popup button now uses the "edit" icon instead of the "add" icon.

![image](https://cloud.githubusercontent.com/assets/912649/21487212/db94ed3e-cb91-11e6-93be-0249a16c9eb9.png)

**Comment by FelisCatus** [comment]
Here's another proposal, which has a flat list of all proxy servers. Mock up:

![image](https://user-images.githubusercontent.com/912649/27007245-eb1d2b1c-4e19-11e7-9d41-e4f234f3cce6.png)

Note that I've also tried to combine the feature request of falling back to DIRECT. Otherwise it is pretty much 1:1 relationship to PAC results (as compared to system proxy settings, Chrome's proxy data structure, etc.).

This feels more natural to set up multiple proxy servers at the expense of harder configuration of Schemes. However, using different proxies on different schemes is a rare case and people often get confused by the old "Show Advanced" settings even if they don't need it. Therefore, scheme is moved to the button there on the right hand side with a popup modal like this:

![image](https://user-images.githubusercontent.com/912649/27007269-c039bab8-4e1a-11e7-9d43-bcba9af4ab35.png)

No functionality removed though, just moved to a different place.

**Comment by loadwiki** [comment]
I have a concern. "proxy servers are tried one by one from top to bottom" for each request sounds bad to performance. Does Chrome cache TCP connections with proxy servers? Is it possible to select the best one (according to ping quality) for each request, and update the selection regularly?

**Comment by celeron533** [comment]
The core of SwitchyOmega is to generate an PAC file and may need to tweak it with user's interaction then regenerate the PAC.
How to compile, use and consume the PAC is controlled by browser itself.

**Comment by olloff** [comment]
As you mentioned before, 
> confusing part about URLs are hidden behind the "Scheme" button.

is probably not the primary confusing part. 

https://github.com/FelisCatus/SwitchyOmega/issues/578#issuecomment-307598067 proposal is great, it probably could save me from made such a novice mistake I made. I'd still add an HTTPS(TLS)-proxy tho but it would never use because of proxy priority. 

It may sound illogical to you, but main confusing part here is PROTOCOL column itself. As a user I bound HTTP proxy protocol with http:// url scheme and HTTPS proxy protocol with https:// url scheme. Here's [how that problem solved in FoxyProxy](https://i.imgur.com/AP1MLX7.png) - to activate HTTPS-proxy user needs to do one additional click and he's informed about SSL proxy with small tooltip.
Best way to treat that behavior is:
- to rename HTTP/HTTPS to something less familiar, such as "regular/HTTP/S proxy" and "secure/SSL/TLS proxy";
- to give user field of thoughts, [returning url schemas back](https://i.imgur.com/jDY6Dl0.png);
- or to imply that proxy-protocol is [not meant for specific url scheme only](https://i.imgur.com/8WLjgS3.png), in this example info is gathered from corresponded switch profile and url-sheme is treated as a pattern.

Btw, thanks for giving me a chance to put my five cent into the frame of discussion. Obviously I'm not a UX-designer, so I can't bring an expert opinion here, but I believe even users like me can be taught by guidance of UI.

**Comment by FelisCatus** [comment]
I agree with you on the protocol part. The technically-right-but-not-so-confusing way would be "HTTP Proxy (Secure)" or "HTTP Proxy over TLS" for these kind of proxies we are talking. They are rare though and usually can't be configured in System Proxy Settings so I there shouldn't be too many users using it, so it makes sense to use some specialized terms to avoid confusing with the URLs. Also, hiding it under a check like FoxyProxy also makes sense. I'm leaning toward the first option now but certainly, a tool tip can be shown regarding what they are and when they are not needed.

For the URL scheme selection, I will propose a third option, as shown below.
![image](https://user-images.githubusercontent.com/912649/29100140-c94b21ce-7c77-11e7-89bc-4e958b313dd3.png)

First, it tries to make what it is explicit by avoiding confusing terms like scheme/protocol, but state clearly what it really is -- Use this proxy for what URLs. It defaults to "All URLs" which make sense in most setups and does not lead to confusing about what checkbox does what. Also, the whole UI design is focused on inputting the proxy itself first -- which is the most important -- and the URL configuration can come later as an additional configuration. If we put protocol and scheme together, there probably can be a moment where the user is thinking, "yeah, I have an HTTP proxy, but do I specify this in the first or second column?", which did lead to mis-configuration as I've seen earlier.

Now if you click on the dropdown, you will see:

![image](https://user-images.githubusercontent.com/912649/29100428-6b04d144-7c79-11e7-9928-a81a4b2cb16d.png)

This will be a multi-select drop-down. You can check whatever URLs you want to use and uncheck whatever you don't want to use. In case the user is confused, there's a help option which will popup a window. In that way we can still give the users all options at one click away, without making each row too  thick.

Everyone, please join the conversation and leave your feedback.

**Comment by FelisCatus** [comment]
This issue could need a little bit feedback before we move on.

**Comment by fuckchinaOK** [comment]
![2018-12-18_145010](https://user-images.githubusercontent.com/41405307/50137311-c6510980-02d5-11e9-86bc-cde7bf1ad8ad.png)
